### DISCUSSION TITLE
[IDEA] near-zero latency multiple (IP) video-streams playing

### DISCUSSION THREAD

**Discussion by MarcoRavich** [forart/HyMPS#17] [Ideas]
(preface: the main purpose of this discussion - following the HyMPS's "philosophy" - is to stimulate the confrontation between devs/maintainers of open projects for the goal of their mutual evolution)

For those who need to display multiple (LAN only) video streams in a grid/wall at the lowest possible latency, here are two open software that we have tested and can recommend:
- @iEvgeny's [CCTV Viewer](https://github.com/iEvgeny/cctv-viewer): best of all for (hardware-like) latencies, Linux-only ![](https://dashboard.snapcraft.io/site_media/appmedia/2020/10/Screenshot_20201007_161122.jpg)
- @tsingsee's [EasyPlayer](https://github.com/tsingsee/EasyPlayer-RTSP): multiplatform, simplest GUI ![](https://camo.githubusercontent.com/cbf922c1c179aec8eece7d77967ad4452b403f60c07d998c407fe751ffc7af7c/687474703a2f2f7777772e6561737964617277696e2e6f72672f6769746875622f696d616765732f65617379706c6179657232303136303930383137313032372e706e67)

It would be great to "combine" these peculiar characteristics to obtain a platform indipendent and easy to use player for live monitoring at lowest-possible latencies,

We've also rounded up many other interesting projects in this (raw, at the moment) list: https://github.com/forart/HyMPS/blob/main/RTSP.md

...what do you think about this ?

**Comment by apple-fritter** [comment]
Hello! Just checking in. I'll give this post a little more thorough of a once-over later, but wanted to indicate my interest in assisting you in the endeavor.

I may be a moment to reply as I will think of good questions to ask that will expedite my understanding of your intended implementation.

Cheers, and have a nice day!

**Comment by MarcoRavich** [comment]
Well, as [discussed in this issue](https://github.com/iEvgeny/cctv-viewer/issues/55) @iEvgeny suggests:
_"the absence of video buffering, which is in every player and introduces a significant delay up to several tens of seconds.
In addition, the new implementation uses the Zero-copy rendering method whenever possible."_

We wonder if someone (like @76technician, @Yash-Garg, @onlyet, @janssen70 and hopefully others) could implement this approach in their projects too.

**Reply by 76technician** [reply]
In fact,my player had output buffer in VideoThread，every rtsp player has buffer to ture the audio and video synchronization.
FFmpeg's rtsp receiver may just support tcp,your image will be destory if you use udp.
So,if you want  lower delay,you need use ffmpeg's decoder and encoder module to build your transport moudle or try to use webrtc.

**Reply by MarcoRavich** [reply]
> every rtsp player has buffer to ture the audio and video synchronization. 
>
...and to manage sync between multiple streams, we guess.

> So,if you want lower delay,you need use ffmpeg's decoder and encoder module to build your transport moudle or try to use webrtc.
>
Then do you think that HW-decoding (such as @bmegli's [hardware-video-streaming libs](https://github.com/bmegli/hardware-video-streaming) or @rigaya's - encoding oriented - [ones](https://github.com/rigaya?tab=repositories)) may help to speedup the whole process ?

Anyway, the "zero-copy video rendering" approach - that @iEvgeny proposed too - sounds suitable to reduce latencies.

Some resources about it that may help:

@OpenVINO-dev-contest's _[single and multi-source video decode and preprocess and inference to show the device surface sharing (zero copy)](https://github.com/OpenVINO-dev-contest/decode-infer-on-GPU)_

@paolobarbolini's [_Simple zero copy* streaming multipart decoder implementation_](https://github.com/paolobarbolini/multiparty)

@gnobal's _[sample of ZeroCopyInputStream/ZeroCopyOutputStream to handle a streaming socket using a pre-allocated buffer](https://github.com/gnobal/protobuf-zero-copy-network-stream-sample)_

[Understanding about Zero copy](https://ducmanhphan.github.io/2020-04-06-Understanding-about-Zero-copy/)

["Zero-copy" video rendering with Mali](https://forum.odroid.com/viewtopic.php?t=21948) discussion (many repos linked in)

[Improving Streaming Application Performance with Zero-copy and Asynchronous TDMS](https://www.ni.com/fr-fr/support/documentation/supplemental/17/improving-streaming-application-performance-with-zero-copy-and-a.html)

[![](https://img.youtube.com/vi/axRklJUExWA/maxresdefault.jpg)](https://youtu.be/GRrZFL3z0aY "Zero-Copy Video Streaming on Embedded Systems the Easy Way - Michael Tretter & Philipp Zabel")

Hope that inspires !

**Reply by MarcoRavich** [reply]
Seems that @alexeyvasilyev's [rtsp-client-android](https://github.com/alexeyvasilyev/rtsp-client-android#readme) implements zero-copy in Android too:
> _Unlike [ExoPlayer](https://github.com/google/ExoPlayer) which also supports RTSP, this library does not make any video buffering. Video frames are shown immidiately when they arrive_

**Comment by apple-fritter** [comment]
Thank you for your understanding. I apologize for the late reply.

I am interested in discussing the implementation of overlay content on the video. Could you please provide a detailed explanation of the information or types of data that you would like to have overlaid?

More specifically, I would like to categorize your various "zones" into classes that can have different sets of attributes. I have a few specific questions in mind: Do you prefer the overlay content to be dynamic or static? Would you like certain elements to update at specific intervals? How would you like to interact with the content in order to make updates?

**Comment by MarcoRavich** [comment]
Thanks for your active partecipation, @apple-fritter !

We use the "multi RTSP-player" software (specifically @iEvgeny's [CCTV Viewer](https://github.com/iEvgeny/cctv-viewer) at the moment) to display live video streams from PTZ cameras to let operators remote control.
[we'll post a video soon to have a better idea of the whole system config]

Android tablets (and [app](https://play.google.com/store/apps/details?id=org.wmo.pantiltzoom)) are used as PTZ controllers for the scope, so they need to match this touch gui with monitoring:
![](https://images-na.ssl-images-amazon.com/images/I/91nRlhJr57L.png)

For this kind of use, operators needs at least viewfinder/safe area markers (transparent pngs overlay ?):
![](https://wpmedia-lib.larryjordan.com/wp-content/uploads/2017/02/SafeZones003.jpg)

-> transparent-background viewfinder [usable file here](https://www.vippng.com/png/full/288-2881081_viewfinder-png.png)

Text labels would be also useful (customizable position, colors and fonts would be great):
![Camera-integrated-in-an-operating-light-Although-being-able-to-preview-the-picture](https://github.com/forart/HyMPS/assets/171307/dbb2a29e-97ec-4285-94a5-207fe49e1fdd)

EDIT: @pataelmo ([PTZCameraController](https://github.com/pataelmo/PTZCameraController) author), @EdwardGoomba ([Hawkeye](https://github.com/EdwardGoomba/hawkeye) one) or @TheColditz (behind [CCTV-Companion](https://github.com/TheColditz/CCTV-Companion)) can contribute with more ideas/needs...

...@berglie's [gstreamer-d3d11-overlay](https://github.com/berglie/gstreamer-d3d11-overlay) or @torreyleonard's [RTSP-Utilities](https://github.com/torreyleonard/RTSP-Utilities) may help/inspires.

Thanks again for your interest in the discussion.

**Reply by apple-fritter** [reply]
It is possible to have an overlaid window that is controlled outside the video production environment and is rendered as a superficial layer with a dynamic and/or transparent background, operated by a dedicated volunteer or perhaps automated processes.

Edit: I look forward to viewing your example video for more ideas. Currently I have a few in mind but need to polish my questions to better fit my evolving understanding of your project.

**Reply by MarcoRavich** [reply]
OK, here's a shot of our full gear setup:
![fullsetup](https://github.com/forart/HyMPS/assets/171307/acd4b96b-1d26-4cd1-9c11-aa295b31e84c)

HW list:
0 - APC [BX650CI-GR](https://www.apc.com/shop/hr/en/products/APC-Back-UPS-650VA-230V-AVR-Schuko-Sockets/P-BX650CI-GR)
1 - Blackmagic [ATEM Mini Extreme ISO](https://www.blackmagicdesign.com/products/atemmini/techspecs/W-APS-18)
2 - Zoom [R16](https://zoomcorp.com/en/us/digital-mixer-multi-track-recorders/multi-track-recorders/r16/)
3 - 8 x VHD-V61 (aka [PT12X-SDI](https://ptzoptics.com/sdi/))
4 - Intel [NUC5i5MYHE](https://ark.intel.com/content/www/us/en/ark/products/84861/intel-nuc-kit-nuc5i5myhe.html)*
5 - AVM [Fritz! Box 4020](https://www.amazon.com/dp/B0114SD27Y)**
6 - ASUS [GX-D1081 V3](https://productz.com/en/asus-gx-d1081-v3/p/K7w4K)
7 - 2 x PowerDsine [3504G](https://www.microsemi.com/documents/powerdsine/datasheets/PD-3504G.pdf)
8 - 8 x 165ft [MEIPEK cat 6 ethernet](https://www.amazon.com/dp/B0936C9XYF) + [Twozoh fiber HDMI](https://www.amazon.com/dp/B0B6ZWX9T6)
9. ASUS [K53S](https://www.techradar.com/reviews/pc-mac/laptops-portable-pcs/laptops-and-netbooks/asus-k53s-1086926/review) (i7-2630QM/8 GB/128SSD)
10. Behringer [MICROAMP HA400](https://www.behringer.com/behringer/product?modelCode=P0386) + [BH470](https://www.behringer.com/product.html?modelCode=P0E3Q)

Note that almost everything (except cables and Behringer stuff) was purchased used - both at Wharehouse and eBay - or is reused.

**Reply by apple-fritter** [reply]
Thank you.

Are you able to clarify the use of a "safe zone?" Is that an area that will be outside of recording? What is safe in the "safe zone?"

Furthermore what in the overlay needs to match the GUI of the ptz controller app, and how would it be useful?

In particular I am having trouble understanding these two considerations.

**Reply by MarcoRavich** [reply]
Thank you so mutch for your interest.

> Are you able to clarify the use of a "safe zone?" Is that an area that will be outside of recording? What is safe in the "safe zone?"
> 
So called "safe areas/zones" are relative to video borders:
https://larryjordan.com/articles/why-do-safe-zones-exist-in-video/
https://en.wikipedia.org/wiki/Safe_area_(television)

Anyway, for a PTZ operator, a viewfinder is essential for panning and zooming "correctness".

However, even if the @tiaanduplessis' one (dynamic on-the-fly generation) is the best approach, from the RTSP-app dev point of view the simplest way is to let users overlay a transparent background image (that matches the displayed stream resolution, of course).

> Furthermore what in the overlay needs to match the GUI of the ptz controller app, and how would it be useful?
> 
We do simply need customizable labels for streams that we can match with controller-app's ones:
![cameras_names](https://github.com/forart/HyMPS/assets/171307/95f92145-e16f-42ea-952c-963ec3aab327)

As before, in this case, the simplest approach should be to let users overlay customizable (position, size, color and - why not - font) texts over displayed streams.

We hope these details will be useful to dispel your doubts about.

Thanks again.

**Reply by apple-fritter** [reply]
> Thank you so mutch for your interest.
> 
> > Are you able to clarify the use of a "safe zone?" Is that an area that will be outside of recording? What is safe in the "safe zone?"
> > 
> So called "safe areas/zones" are relative to video borders:
> https://larryjordan.com/articles/why-do-safe-zones-exist-in-video/
> https://en.wikipedia.org/wiki/Safe_area_(television)


I see! You're ensuring that the subject remains in the frame, even in the event of movement, and you're assigning space to account for overlay that isn't disruptive or distracting of the content.

So I believe first we should determine the design of the viewfinder as it faces the operator/viewer, then I can work on the implementation of rendering it, and the strategies to correlate and control the overlay content to the viewfinder design.

**Reply by MarcoRavich** [reply]
> So I believe first we should determine the design of the viewfinder as it faces the operator/viewer, then I can work on the implementation of rendering it, and the strategies to correlate and control the overlay content to the viewfinder design.
>
Thanks for the deep interest, but these choices must be customizable and not pre-setted (every operator has his own needs/preferences about viewfined look&feel), IOHO.
That's why we believe in the "less is more" approach: simply let users to select a PNG image (with a transparent background, of course) and just overlay it to the displayed stream. 
That'a all (or, well, in case of mismatching image-to-stream resolution, an on-the-fly scaling may be needed).

Here are some interesting FFMPEG-based articles/discussions about image overlaying that may inspire:
- https://creatomate.com/blog/how-to-add-a-transparent-overlay-on-a-video-using-ffmpeg
- https://www.abyssale.com/generate-video/ffmpeg-overlay-image-on-video
- https://video.stackexchange.com/questions/34284/overlay-transparent-image-png-over-video-mp4-with-ffpmeg
- https://superuser.com/questions/868204/overlay-transparent-animation-over-video-with-ffmpeg
- https://stackoverflow.com/questions/38734068/ffmpeg-make-areas-transparent-before-overlaying-a-vid-with-perspective
- https://stackoverflow.com/questions/69536515/overlay-image-over-a-transparent-gif-using-ffmpeg
- https://stackoverflow.com/questions/10918907/how-to-add-transparent-watermark-in-center-of-a-video-with-ffmpeg

Anyway we suggest to use [this file](https://www.vippng.com/png/full/288-2881081_viewfinder-png.png) for testing, if you need.

Last but not least, you can "grab" text overlay methods from @jb-alvarado's [ffplayout](https://ffplayout.github.io/), that offers a very clean UI ([ffplayout-frontend](https://github.com/ffplayout/ffplayout-frontend#readme)) too:
![](https://raw.githubusercontent.com/ffplayout/ffplayout-frontend/master/docs/images/message.png)

Hope that helps.

**Reply by apple-fritter** [reply]
That will be far easier than I pictured you to need.

I will view the project later today and see if I can come up with a set of possible solutions with fresh code.

I have some potential queries for you:
Can you clarify the relationship of the PTZ controls to the overlay content?
Am I conflating different topics?
As the viewfinder is needing to be customizable in design, will the user's preference in design also impact the boundaries for other overlay content (titles, labels, etc)... that is to say, will it adjust the definition of safe zones?

EDIT: I feel a vector image might be a good consideration for the viewfinder as it would render cleanly irrespective of scaling performed.

**Reply by MarcoRavich** [reply]
( note: images and text overlays should be - of course - treated as different topics so, from now on, we will use "topic tag" before the relative part )

## TEXTS OVERLAY

> Can you clarify the relationship of the PTZ controls to the overlay content?
>
Since each PTZ operator moves one camera at a time, the controllers (both hw and sw) they use allows to select which one.
Typically "nametags" are chosen to identify each remote camera ('cause simplifies their management rather than IPs), so these labels must - of course - match both on controllers and monitors.
(we'll post the - infamous - video about this soon: it will be more explanatory than words)

For this purpose we believe that text(s) overlay - therefore editable on the fly - should be the best choice:

[![](https://img.youtube.com/vi/XNuFAE35mMo/hqdefault.jpg)](https://youtu.be/XNuFAE35mMo "[Youtube] How to render text onto a video using ffmpeg") 

[![](https://img.youtube.com/vi/2X4yqynOaTU/hqdefault.jpg)](https://youtu.be/2X4yqynOaTU "[Youtube] FFMPEG Overlay And Drawtext: How To Easily Render Text And Images On Your Video!")

- https://hhsprings.bitbucket.io/docs/programming/examples/ffmpeg/drawing_texts/drawtext.html
- https://ottverse.com/ffmpeg-drawtext-filter-dynamic-overlays-timecode-scrolling-text-credits/
- https://transkriptor.com/add-text-to-video-ffmpeg/
- https://stackoverflow.com/questions/54789728/ffmpeg-command-with-overlay-and-drawtext-on-video
- https://superuser.com/questions/939357/how-to-position-drawtext-text
- https://stackoverflow.com/questions/49454740/multiple-overlays-using-ffmpeg
- https://video.stackexchange.com/questions/25574/realtime-time-overlay-with-ffmpeg


## IMAGE OVERLAY

We believe that viewfinder (= trasparent-background image) shape/color/thickness/etc should be a sole users' burden: their widest freedom of choice must be retained as much as possible.

For example they could decide to BUY one (or even more) [online](https://pngtree.com/so/viewfinder)...
![VFbuy](https://github.com/forart/HyMPS/assets/171307/9a87dfc5-a3ea-4ea9-88dd-5da6122b112d)

....or, why not, choose to use a (provocatory - even if perfectly usable - of course) kid-freehand-drawned one:
![VFmine](https://github.com/forart/HyMPS/assets/171307/f3aa9764-cc38-4a87-b4f8-266bee6a761f)

To us RTSP display software must be limited to render the overlay after an image-to-stream resolution adapting (better, of course. if SVG vector files are provided).

It would be mutch useful to let users choose a different viewfinder (= trasparent-background image) for each stream.

In the end yes, [KISS](https://en.wikipedia.org/wiki/KISS_principle) is the key.

**Reply by MarcoRavich** [reply]
OK, we've digged a little more and discovered that a basic one could be "drawable" directly by FFMPEG (using _[drawgrid](https://ffmpeg.org/ffmpeg-filters.html#drawgrid)_ and/or _[drawbox](https://ffmpeg.org/ffmpeg-filters.html#drawbox)_ commands):

[![](https://img.youtube.com/vi/gl0bS1Gr7dc/hqdefault.jpg)](https://youtu.be/gl0bS1Gr7dc "[Youtube] Examples of visualization of video data with ffmpeg - drawgrid") 
Source: https://hhsprings.bitbucket.io/docs/programming/examples/ffmpeg/video_data_visualization/drawgrid.html

[![](https://img.youtube.com/vi/y8-T_pEMe3o/hqdefault.jpg)](https://youtu.be/y8-T_pEMe3o "[Youtube] Examples of visualization of video data with ffmpeg - drawbox") 
Source: https://hhsprings.bitbucket.io/docs/programming/examples/ffmpeg/video_data_visualization/drawbox.html

...perhaps @afrum could help to build a right commandline to draw "[EBU-compliant video safe areas](https://tech.ebu.ch/docs/r/r095.pdf)" (and, possibly, a rudimentary viewfinder) since he has already made [several interesting scripts](https://github.com/afrum/my_ffmpeg)...

We've also opened a question both @ [SO](https://stackoverflow.com/questions/76488249/how-to-draw-dynamic-safe-areas-in-ffplay) / [SE](https://video.stackexchange.com/questions/36588/how-to-draw-scalable-safe-areas-in-ffplay) where we posted this (1080p) scheme too:
![1080p_safe](https://github.com/forart/HyMPS/assets/171307/797b81f0-3b99-4a15-bc60-c175603390a4)

**Reply by MarcoRavich** [reply]
Here we go, (driven) ChatGPT did it:
`ffplay -vf "drawgrid=x=(iw/2):(ih/2):t=3:color=yellow@0.5, drawbox=x=(iw-iw*0.95)/2:y=(ih-ih*0.95)/2:w=iw*0.95:h=ih*0.95:color=red@1, drawbox=x=(iw-iw*0.965)/2:y=(ih-ih*0.965)/2:w=iw*0.965:h=ih*0.965:color=green@1" filename.ext`

**Reply by MarcoRavich** [reply]
Ok, after a quite long chat, here's the updated commandline (with added 4:3 safe area too):

`ffplay -vf "drawgrid=x=(iw/2):(ih/2):t=3:color=yellow@0.5, drawbox=x=(iw-(iw*0.675))/2:y=(ih-ih)/2:w=iw*0.675:h=ih:color=black@0.3:t=fill, drawbox=x=(iw-iw*0.95)/2:y=(ih-ih*0.95)/2:w=iw*0.95:h=ih*0.95:color=red@1, drawbox=x=(iw-iw*0.965)/2:y=(ih-ih*0.965)/2:w=iw*0.965:h=ih*0.965:color=green@1" filename.ext`

Hope that inspires.

**Reply by MarcoRavich** [reply]
> Can you clarify the relationship of the PTZ controls to the overlay content?

Here's te video (encoded by Canon HF100 camcorder @ 17 MBps):
http://www.forart.it/_priv/PTZ_ctrl-view.mp4

The operator(s) needs to clearly identify the controller-selected camera in preview monitoring(s) to avoid possible mess and/or misunderstanding with the director.
That's why (customizable) labels are needed.

Hope it helps.

**Comment by MarcoRavich** [comment]
Just found a very interesting - advanced - approach for *generating* a viewfinder (but needs to be adapted, eventually):
@tiaanduplessis's [react-native-view-finder](https://github.com/tiaanduplessis/react-native-view-finder)

[Usage](https://github.com/tiaanduplessis/react-native-view-finder#usage) example:
```jsx
import {View, StyleSheet} from 'react-native'
import ViewFinder from 'react-native-view-finder'

const Example = () => (
  <View style={styles.container}>
    <ViewFinder backgroundColor="#333" loading />
  </View>
)

const styles = StyleSheet.create({
  container: { flex: 1 }
})
```

This will result in:
![](https://raw.githubusercontent.com/tiaanduplessis/react-native-view-finder/master/example.png)

Hope that inspires.

**Reply by apple-fritter** [reply]
That is an interesting project. If you enjoy the design, that should be easy to accomplish. I'll visit the code tomorrow to see how the strategies used could be transmuted to fit your purposes.

**Comment by MarcoRavich** [comment]
>     * @tsingsee's [EasyPlayer](https://github.com/tsingsee/EasyPlayer-RTSP): multiplatform, simplest GUI ![](https://camo.githubusercontent.com/cbf922c1c179aec8eece7d77967ad4452b403f60c07d998c407fe751ffc7af7c/687474703a2f2f7777772e6561737964617277696e2e6f72672f6769746875622f696d616765732f65617379706c6179657232303136303930383137313032372e706e67)
>
We've just realized that [EasyPlayer](https://github.com/tsingsee/EasyPlayer-RTSP)' screenshots are no longer reachable in repos managed by @babosa, sorry !

EDIT:
Well, we hope other RTSP-streams "grid player" projects' devs (such as @grigory-lobkov, @yuriizubkov and @Mawkey just to cite some) are somewhat interested in this discussion.

**Reply by yuriizubkov** [reply]
Hi folks!
I'm not sure how much help I can be. I created my "grid player" script specifically for using on Raspberry Pi (about 6 years ago). Script simply controls omxplayer instances (because omxplayer has hardware acceleration for video playback on the Rpi). I don't even know if it works on other platforms. :)

**Reply by apple-fritter** [reply]
> Hi folks!
> I'm not sure how much help I can be. I created my "grid player" script specifically for using on Raspberry Pi (about 6 years ago). Script simply controls omxplayer instances (because omxplayer has hardware acceleration for video playback on the Rpi). I don't even know if it works on other platforms. :)

Hi Yuri! I looked at your video wall project... Interesting. I will revisit it later when I have more time. I have a couple RTSP repos up and will be creating my own meta sections and repo for them in the coming days, if you are interested.

**Reply by Tom722620** [reply]
work

**Comment by MarcoRavich** [comment]
BUMP

Just discovered @illuspas's [NodeMedia Clients](https://github.com/NodeMedia/) (Android and iOS) which implements:

- Hardware Acceleration
- Low latency
- Delay elimination 

Both for [play](https://github.com/NodeMedia/NodeMediaClient-Android#play) and [publish](https://github.com/NodeMedia/NodeMediaClient-Android#publish).

**Comment by bitvilag** [comment]
I landed here as https://github.com/Anonymousdog/displaycameras is not supported on Bullseye.
As far as I can see it is related to the missing live555 from debian due to license issues. (I really dont understand why these companies are this needy in this matter).

The project uses omxplayer, but that is deprecated and as far as I can se is deprecated because of the missing package.

What do you see as the closes implementation to this project? I would need a couple rtsp feeds on my RP1 to be presented on the attached monitor.

**Comment by MarcoRavich** [comment]
> What do you see as the closes implementation to this project? I would need a couple rtsp feeds on my RP1 to be presented on the attached monitor.
>
Well, at the moment the best - and most updated - RTSP "player" is [CCTV Viewer](https://github.com/iEvgeny/cctv-viewer#cctv-viewer) without any doubt.

Anyway it runs on a "standard" linux distro, dunno if works on RPiOS.

Try it yourself and report to @iEvgeny, please.

**Reply by bitvilag** [reply]
I saw somewhere that if I go back to Buster I am good for a while as its suppose to be supported until mid of 2024.
I checked the project and at first the README did not really encourge me thats something easy to setup. 

I will check back later when I guess more docs are present, but for now I will try to stick with Buster and omxplayer

**Reply by MarcoRavich** [reply]
> I checked the project and at first the README did not really encourge me thats something easy to setup.
> 
Well, if you search between [CCTV Viewer issues](https://github.com/iEvgeny/cctv-viewer/issues?q=), you can find some interesting infos...
- [How to compile from source on arm/raspberry PI](https://github.com/iEvgeny/cctv-viewer/issues/34)

@dvision1979 seems the most experienced user about...
